src/intentManager.js

```diff
@@ -51,167 +55,98 @@ define([
-                        xmlWriter, intentMug.p.nodeID);
-                });
-            }
-        };
-        return that;
-    };
-    
-    function androidIntentAppId(mug, options) {
-        options.id = "intent-app-id";
-        var widget = widgets.base(mug, options);
+            unmappedIntentTags[tagId] = newTag;
+        });
+    }
 
-        widget.definition = mug.p.getDefinition('androidIntentAppId');
-        widget.currentValue = (mug.intentTag) ? mug.intentTag.getAttr('path') : "";
-        
-        var input = $("<input />")
-            .attr("name", widget.id)
-            .attr("type", "text")
-            .attr('placeholder', 'Insert Android Application ID');
+    function syncMugWithIntent (mug) {
+        // called when initializing a mug from a parsed form
+        if (mug.__className === "AndroidIntent") {
+            var nodeID = mug.p.nodeID,
+                tag = _.findWhere(unmappedIntentTags, {nodeID: nodeID});
 
-        widget.getControl = function () {
-            if (widget.isDisabled()) {
-                input.prop('disabled', true);
+            if (!tag) {
+                tag = makeODKXIntentTag(nodeID, null);
             }
-            return input;
-        };
-
-        widget.setValue = function (value) {
-            input.val(value);
-        };
 
-        widget.getValue = function() {
-            return input.val();
-        };
-
-        widget.updateValue = function () {
-            widget.mug.intentTag.setAttr('path', widget.getValue());
-        };
+            _.each(INTENT_SPECIFIC_SPECS, function (key) {
+                mug.p[key] = tag[key];
+            });
 
-        input.bind("change keyup", widget.updateValue);
-        return widget;
+            delete unmappedIntentTags[tag.nodeID];
+        }
     }
 
-    function androidIntentExtra(mug, options) {
-        options.id = "intent-extra";
-        var widget = widgets.baseKeyValue(mug, options);
-        widget.currentValue = (mug.intentTag) ? mug.intentTag.getAttr('extra') : {};
-        widget.definition = mug.p.getDefinition('androidIntentExtra');
+    function writeIntentXML (xmlWriter, tree) {
+        // make sure any leftover intent tags are still kept
+        _.each(unmappedIntentTags, function (tag) {
+            writeXML(xmlWriter, tag);
+        });
 
-        widget.save = function () {
-            if (widget.mug.intentTag) {
-                widget.mug.intentTag.setAttr('extra', widget.getValidValues());
+        var intents,
+        getIntentMugs = function(node) {
+            var mug = node.getValue();
+            if (!mug || node.isRootNode) {
+                return null;
+            }
+            if (mug.options.dataType === 'intent') {
+                return mug;
+            } else {
+                return null;
             }
         };
+        intents = tree.treeMap(getIntentMugs);
+        intents.map(function (intentMug) {
+            writeXML(xmlWriter, intentMug.p);
```

Comment:
Would it work to move this line up to [line 128](https://github.com/dimagi/Vellum/pull/497/files#diff-ebaaf3f85de0ada3835858976fdf5ba0R128)? Then you could ignore the `treeMap` return value.